File: 04 Sorting Algorithm/Sorting Algorithm - customized comparator.cpp
#include <bits/stdc++.h>
using namespace std;
class Student {
public:
    string name;
    int grade, age;
    double score;
    Student(){};
    Student(string name, int grade, int age, double score){
        this->name = name;  this->grade = grade; this->age = age; this->score = score;
    }
};
bool agecmp(Student a, Student b){// return true (a before b), false (a after b)
    if(a.age < b.age) return true;
    // if(a.age == b.age) return(a.grade < b.grade); // Additional sort
    return false;
}
int main(){
    int n; vector<Student> vec;
    cin >> n;
    for(int i=0; i<n; i++){
        string name; int grade, age; double score;
        cin >> name >> grade >> age >> score;
        vec.push_back(Student(name, grade, age, score));
    }
    sort(vec.begin(), vec.end(), agecmp);
    for(Student x : vec) {
        cout << "name:" << x.name << " grade: " << x.grade << " age:" << x.age << " score:" << x.score << endl;
    }
}
/*
3
Alice 10 15 96.5
Bob 10 14 93.8
Calvin 9 14 97.5
*/

File: 04 Sorting Algorithm/Sorting Algorithm - customized comparator.java
import java.util.*;
public class Main {
	public static void main(String[] args) {
		Scanner in = new Scanner(System.in);
		int n = in.nextInt();
		student arr [] = new student[n];
		for(int i=0; i<n; i++) {
			String name = in.next(); 
			int grade = in.nextInt(), age = in.nextInt();
			double score = in.nextDouble();
			arr[i] = new student(name, grade, age, score);
		}
		Arrays.sort(arr, new ageComparator());
		for(int i=0; i<n; i++) {
			arr[i].print();
		}
	}
	static class ageComparator implements Comparator<student>{
		public int compare(student a, student b) {
			return Integer.compare(a.age, b.age);
		}
	}
	static class student {
		static String school;
		String name;
		int grade, age;
		double score;
		student(){ };
		student(String name, int grade, int age, double score) { 
			this.name = name; this.grade = grade; this.age = age; this.score = score; 
		}
		public void print() { System.out.println("name: " + name + " grade: " + grade + " age: " + age); }
	}
}
/*
3
Alice 10 15 96.5
Bob 10 14 93.8
Calvin 9 14 92.5
*/
